viewer/viewer.js

```diff
@@ -7686,24 +7686,126 @@ function buildHuntOptions (huntId, hunt) {
   return options;
 }
 
+// if we couldn't retrieve the seession, skip it but add it to failedSessionIds
+// so that we can go back and search for it at the end of the hunt
+function continueHuntSkipSession (hunt, huntId, session, sessionId, searchedSessions, cb) {
+  if (!hunt.failedSessionIds) {
+    hunt.failedSessionIds = [ sessionId ];
+  } else {
+    // pause the hunt if there are more than 10k failed sessions
+    if (hunt.failedSessionIds.length > 10000) {
+      return pauseHuntJobWithError(huntId, hunt, {
+        value: 'Error hunting: Too many sessions are unreachable. Please contact your administrator.'
+      });
+    }
+    // make sure the session id is not already in the array
+    // if it's not the first pass and a node is still down, this could be a duplicate
+    if (hunt.failedSessionIds.indexOf(sessionId) < 0) {
+      hunt.failedSessionIds.push(sessionId);
+    }
+  }
+
+  return updateHuntStats(hunt, huntId, session, searchedSessions, cb);
+}
+
+// if there are failed sessions, go through them one by one and do a packet search
+// if there are no failed sessions left at the end then the hunt is done
+// if there are still failed sessions, but some sessions were searched during the last pass, try again
+// if there are still failed sessions, but no new sessions coudl be searched, pause the job with an error
+function huntFailedSessions (hunt, huntId, options, searchedSessions, user) {
+  let changesSearchingFailedSessions = false;
+  if (hunt.failedSessionIds && hunt.failedSessionIds.length) {
+    options.searchingFailedSessions = true;
+    // copy the failed session ids so we can remove them from the hunt
+    // but still loop through them iteratively
+    let failedSessions = JSON.parse(JSON.stringify(hunt.failedSessionIds));
+    // we don't need to search the db for session, we just need to search each session in failedSessionIds
+    async.forEachLimit(failedSessions, 3, function (sessionId, cb) {
+      Db.get(Db.sid2Index(sessionId), 'session', Db.sid2Id(sessionId))
```

Comment:
this should use Db.getSession if possible

Reply:
Does Db.getSession get the session even if it's on a remote node? So I won't have to do the make request part too?

Reply:
or do something like this:

  ```
Db.getSession(sessionId, {}, (err, session) => {
      if (err) {
        return continueHuntSkipSession(hunt, huntId, session, sessionId, searchedSessions, cb); 
      }
      
      makeRequest(session.node, `${session.node}/hunt/${huntId}/remote/${sessionId}`, user, (err, response) => {
        if (err) {
          return continueHuntSkipSession(hunt, huntId, session, sessionId, searchedSessions, cb);
        }

        let json = JSON.parse(response);

        if (json.error) {
          console.log(`Error hunting on remote viewer: ${json.error} - ${path}`);
          return continueHuntSkipSession(hunt, huntId, session, sessionId, searchedSessions, cb);
        }

        // remove from failedSessionIds if it was found
        hunt.failedSessionIds.splice(hunt.failedSessionIds.indexOf(sessionId), 1);
        // there were changes to this hunt; we're making progress
        changesSearchingFailedSessions = true;

        if (json.matched) { hunt.matchedSessions++; }
        return updateHuntStats(hunt, huntId, session, searchedSessions, cb);
      });
    });
```

Reply:
And I don't need to pass it options because I don't need any fields. I just need the packets. Right?

Reply:
Oh i just realized it doesn't have a promise version :( And yes you can just do {} for options, although I think you just need the ['lastPacket', 'node', 'huntId', 'huntName', 'fileId'] fields.

Reply:
👍 works perfectly. I forgot that function existed!